src/Storages/FileLog/StorageFileLog.cpp

```diff
@@ -0,0 +1,746 @@
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <IO/ReadBufferFromFile.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteBufferFromFile.h>
+#include <IO/WriteIntText.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/InterpreterInsertQuery.h>
+#include <Interpreters/evaluateConstantExpression.h>
+#include <Parsers/ASTCreateQuery.h>
+#include <Parsers/ASTExpressionList.h>
+#include <Parsers/ASTInsertQuery.h>
+#include <Parsers/ASTLiteral.h>
+#include <Processors/Executors/CompletedPipelineExecutor.h>
+#include <Processors/Pipe.h>
+#include <Processors/Transforms/ExpressionTransform.h>
+#include <Storages/FileLog/FileLogSource.h>
+#include <Storages/FileLog/ReadBufferFromFileLog.h>
+#include <Storages/FileLog/StorageFileLog.h>
+#include <Storages/StorageFactory.h>
+#include <Storages/StorageMaterializedView.h>
+#include <Poco/File.h>
+#include <Common/Exception.h>
+#include <Common/Macros.h>
+#include <Common/getNumberOfPhysicalCPUCores.h>
+#include <Common/quoteString.h>
+#include <Common/typeid_cast.h>
+#include <common/logger_useful.h>
+
+#include <sys/stat.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int BAD_ARGUMENTS;
+    extern const int CANNOT_GET_FILE_STAT;
+    extern const int NOT_REGULAR_FILE;
+    extern const int READ_META_FILE_FAILED;
+    extern const int FILE_STREAM_ERROR;
+    extern const int LOGICAL_ERROR;
+}
+
+namespace
+{
+    const auto RESCHEDULE_MS = 500;
+    const auto BACKOFF_TRESHOLD = 32000;
+    const auto MAX_THREAD_WORK_DURATION_MS = 60000;
+}
+
+StorageFileLog::StorageFileLog(
+    const StorageID & table_id_,
+    ContextPtr context_,
+    const ColumnsDescription & columns_,
+    const String & relative_path_,
+    const String & format_name_,
+    std::unique_ptr<FileLogSettings> settings,
+    bool attach)
+    : IStorage(table_id_)
+    , WithContext(context_->getGlobalContext())
+    , filelog_settings(std::move(settings))
+    , path(getContext()->getUserFilesPath() + "/" + relative_path_)
```

Comment:
So now it is assumed that `path` in storage arguments is relative to `user_files_path`? But, it is incorrect, it does not have to be relative. Need to canonize it and compare to `user_files_path` and throw exception if canonized path does not start with another canonized path.
And also if we want to allow symlinks need to use `symlinkStartsWIth` from fileSystemHelpers.cpp.

Reply:
I'm not very clear about it, so the path arguments should be a absolute path?

Reply:
No, it can be any path, relative, absolute, not-normal. But we canonize it and check with canonized `user_files_path` (better use symlinkStartsWIth method, it will normalize both paths itself).

Reply:
But if it is a relative path, such as `a/`, we should treat it relative to `user_files_path`?

Reply:
No, relative to std::filesystem::current_path() -- path where server was started. But it should not matter because we make it absolute and only then compare.

Reply:
got it.